feature/Diagram validation tool

mrmasterplan opened the pull request:
# Diagram Validation

```
$> spetlr-check-diagram                                                      
usage: spetlr-check-diagram [-h] [--tag TAG] module diagram
spetlr-check-diagram: error: the following arguments are required: module, diagram
```

The assumed setup is that you have a python library, say `dataplatform`, that is 
importable in the local virtual environment. Individual classes in the 
`dataplatform` library can then be marked up with docstrings as follows.
Example `src/dataplatform/my/folder/SomeOrchestrator.py`:
```python
class SomeOrchestrator:
    """
    Some free text describing my orchestrator.
    ```spetlr diagram
    node:
        name: SomeOrchestrator
        incoming:
            - SourceTable
            - DetailsTable
        outgoing:
            - ResultsTable
    ```
    """
```

This tool works only on drawio diagrams for now. It scans the drawio diagram for 
arrows that connect any two entities that are connected with a line, registering a 
source to target connection for each line. The type of symbol, or style of line are 
irrelevant. Suppose you have a diagram as in the following example.

![diagram_eample.png](docs/cli_tools/diagram_eample.png)

You can now run
```powershell
$> spetlr-check-diagram dataplatform docs/my-dataplatform-diagram.drawio
```

The tool will tell you that you are missing a connection from DetailsTable to 
SomeOrchestrator and will return an error code. If you add such an arrow and rerun 
the tool, the command will return without exit code.

LauJohansson commented on src/spetlrtools/diagrams/DiagramParser.py:
```diff
@@ -0,0 +1,77 @@
+                continue
+        else:
+            self.label = ""
+
```
If the loop hits the `except KeyError: continue` - will it jump down to the else statement afterwards?

(I try to figure out if there is a possibility that self.label is not setted.

mrmasterplan replied:
I think it works, but I changed it to make it clearer.

LauJohansson commented on src/spetlrtools/diagrams/Edge.py:
```diff
@@ -0,0 +1,10 @@
+_Edge = namedtuple("_Edge", "source target")
+
+
+class Edge(_Edge):
```
why no __init__?

LauJohansson replied:
add docstring

LauJohansson commented on src/spetlrtools/diagrams/Edge.py:
```diff
@@ -0,0 +1,10 @@
+
+
+class Edge(_Edge):
+    def __new__(cls, source: str, target: str, *, style=None):
```
add return type

mrmasterplan replied:
__new__ returns an uninitalized blank object that is fed into __init__. I don't think it needs a return type since it is never used expicitly.

LauJohansson commented on src/spetlrtools/diagrams/DiagramParser.py:
```diff
@@ -0,0 +1,77 @@
+        self.edges: List[DiagramEdge] = []
+        self.nodes: Dict[str, DiagramNode] = {}
+
+    def parse(self):
```
add return type?

LauJohansson commented on src/spetlrtools/diagrams/HTMLStripper.py:
```diff
@@ -0,0 +1,24 @@
+        self.convert_charrefs = True
+        self.text = StringIO()
+
+    def handle_data(self, d):
```
add return type

LauJohansson commented on src/spetlrtools/diagrams/HTMLStripper.py:
```diff
@@ -0,0 +1,24 @@
+        self.text.write(d)
+        return self
+
+    def get_data(self):
```
add return type

LauJohansson commented on src/spetlrtools/diagrams/HTMLStripper.py:
```diff
@@ -0,0 +1,24 @@
+        return self.text.getvalue()
+
+    @staticmethod
+    def strip(text):
```
add return type

LauJohansson commented on src/spetlrtools/diagrams/HTMLStripper.py:
```diff
@@ -0,0 +1,24 @@
+from io import StringIO
+
+
+class HTMLStripper(HTMLParser):
```
add docstring

LauJohansson commented on src/spetlrtools/diagrams/DiagramParser.py:
```diff
@@ -0,0 +1,77 @@
+        return Edge(self.source_label, self.target_label)
+
+
+class DiagramParser:
```
add docstring

LauJohansson commented on src/spetlrtools/diagrams/DiagramParser.py:
```diff
@@ -0,0 +1,77 @@
+            self.label = ""
+
+
+class DiagramEdge(DiagramNode):
```
add docstring

LauJohansson commented on src/spetlrtools/diagrams/DiagramParser.py:
```diff
@@ -0,0 +1,77 @@
+from spetlrtools.diagrams.HTMLStripper import HTMLStripper
+
+
+class DiagramNode:
```
add docstring

LauJohansson commented on src/spetlrtools/diagrams/LibraryParser.py:
```diff
@@ -0,0 +1,80 @@
+from spetlrtools.diagrams.Edge import Edge
+
+
+class LibraryParser:
```
Give a more spetlr like name? 

LibraryParser is very generic. 

Like: SpetlrDiagramLibraryParser. Maybe you can find a more readable name

LauJohansson commented on docs/cli_tools/diagram_eample.png:
there is a misspelling in the file name

LauJohansson commented on src/spetlrtools/diagrams/LibraryParser.py:
```diff
@@ -0,0 +1,80 @@
+    def __init__(self, package_name):
+        self.package_name = package_name
+
+    def import_submodules(self, package_name):
```
docstring

LauJohansson commented on src/spetlrtools/diagrams/LibraryParser.py:
```diff
@@ -0,0 +1,80 @@
+                results.update(self.import_submodules(full_name))
+        return results
+
+    def get_class_docstrings_in_module(self):
```
docstring

LauJohansson commented on src/spetlrtools/diagrams/LibraryParser.py:
```diff
@@ -0,0 +1,80 @@
+                        continue
+                    yield name, doc
+
+    def get_relations(self):
```
docstring

LauJohansson commented on src/spetlrtools/diagrams/LibraryParser.py:
```diff
@@ -0,0 +1,80 @@
+                            raise NotImplementedError("styles not supported yet")
+                    for target in node.get("outgoing", []):
+                        if isinstance(target, str):
+                            yield Edge(nodename, target)
```
line 77 and 73 has some discrepancy. Is it on purpose that 73 is `Edge(source, nodename)` and 77 is the opposite `Edge(nodename, target)`?

mrmasterplan replied:
I added a comment

LauJohansson commented on src/spetlrtools/diagrams/main.py:
```diff
@@ -0,0 +1,44 @@
+from spetlrtools.diagrams.LibraryParser import LibraryParser
+
+
+def process_diagram(
```
docstring

LauJohansson replied:
return type

LauJohansson commented on src/spetlrtools/diagrams/main.py:
```diff
@@ -0,0 +1,44 @@
+
+
+def process_diagram(
+    diagram_path: str, module_name: str, tag_name: str = "spetrl diagram"
```
misspelling: "spetlr diagram"

LauJohansson commented on src/spetlrtools/diagrams/main.py:
```diff
@@ -0,0 +1,44 @@
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Compare DrawIO digram edges to docstrings."
```
Mispelling: "Compare DrawIO diagram edges to docstrings."

mrmasterplan commented:
Thank you @LauJohansson for your review. I have
- incorporated all your comments
- added docstrings and return types everywhere that this made sense
- rewritten substantial parts of the feature.
- added support for the other file formats I had mentioned.
- substantially expanded the documentation

mrmasterplan commented:
Pipeline is currently blocked by dependency conflict, waiting for at fix to this issue: [https://github.com/databricks/databricks-cli/issues/644](https://github.com/databricks/databricks-cli/issues/644)
